### ISSUE TITLE
关于微调参数“tempeture”等

### ISSUE THREAD

**Issue by Yyy11181** [FlagOpen/FlagEmbedding#58]
您好！ 在单个gpu上微调遇到几个问题咨询一下您：
1、不知道为什啥 模型训练过程自己这个参数自动变成了这样
08/24/2023 21:41:41 - INFO - FlagEmbedding.baai_general_embedding.finetune.modeling -   reset temperature = 1.0 due to using inner product to compute similarity 我该如何恢复维初始设置的0.01呢。此外temperature: the similarity will be simi = simi/temperature before using them to compute loss. A higher temperature can reduce the value of similarity between texts in downstream tasks.
这个参数不是很理解。能通俗的解释一下吗？
2、到哪里设置为非分布式训练呢？我仅仅在单个gpu上进行微调，一直没找到那个参数
导致单个训练gpu训练：  Process rank: 0, device: cuda:0, n_gpu: 1, **_distributed training: True_,** 16-bits training: False
3、train_group_size 这个参数有点不确定怎么设。我的每条数据里面。pos 和neg 的数量不是确定的，有时候1条，有时候大于2条等，我该如何设置这个参数呢？
期待得到您的回复！感谢

**Comment by staoxiao** [comment]
您好，
1. 报错这个warn是因为没有设置`--normlized True`，这样的化是通过内积计算相似度，是不需要设置温度系数的。只有使用余弦相似度时，才需要温度系数，因为余弦相似度的取值范围非常小，模型很难优化。
2. 但gpu训练，只需要`torchrun --nproc_per_node {number of gpus} `中的{number of gpus}设为1就好
3. train_group_size是控制每次训练时neg的数量，可以取最少neg数量+1或者取平均neg数量+1，+1是因为会使用1个pos。

**Comment by Yyy11181** [comment]
此外，在运行的时候，
![image](https://github.com/FlagOpen/FlagEmbedding/assets/96362397/43531b99-3bea-4598-abf6-1cbe67441f47)
这个normlized=True 参数我代码一直设置不了 ，不知道咋回事，报错是代码问题。

**Comment by staoxiao** [comment]
> 此外，在运行的时候， ![image](https://user-images.githubusercontent.com/96362397/262998718-43531b99-3bea-4598-abf6-1cbe67441f47.png) 这个normlized=True 参数我代码一直设置不了 ，不知道咋回事，报错是代码问题。

不太理解什么意思，可以提供一些更详细的信息和运行的命令

**Comment by Yyy11181** [comment]
我之前把报错保留下来 就是提醒我没有这个参数normlized，所以我就删掉了 ，刚刚加上去了跑了一下，系统自动忽略了这个参数，还是提醒
08/24/2023 22:21:23 - INFO - FlagEmbedding.baai_general_embedding.finetune.modeling -   reset temperature = 1.0 due to using inner product to compute similarity  （无奈）

**Comment by staoxiao** [comment]
3.10的python应该没有问题的，可能是执行命令不对，可以提供一下训练命令查看一下

**Comment by Yyy11181** [comment]
好的，我后续再更新哈，到时候回复您，再次感谢您



---原始邮件---
发件人: ***@***.***&gt;
发送时间: 2023年8月24日(周四) 晚上11:18
收件人: ***@***.***&gt;;
抄送: ***@***.******@***.***&gt;;
主题: Re: [FlagOpen/FlagEmbedding] 关于微调参数“tempeture”等 (Issue #58)




 
3.10的python应该没有问题的，可能是执行命令不对，可以提供一下训练命令查看一下

**Comment by Yyy11181** [comment]
<img width="941" alt="image" src="https://github.com/FlagOpen/FlagEmbedding/assets/96362397/41ab04bf-e6fc-4e58-828d-d9de4147b52d">
 hello,想请教一下您这个参数我设置了不。我是这样设置的：
 
<img width="509" alt="image" src="https://github.com/FlagOpen/FlagEmbedding/assets/96362397/b7b13cc8-6b7e-49ad-85c5-5d2879989419">
但是报错：
<img width="1004" alt="image" src="https://github.com/FlagOpen/FlagEmbedding/assets/96362397/0e2f214c-6327-4d41-825d-40b731b0774c">

**Comment by staoxiao** [comment]
这个参数目前没有加入到训练脚本里，如果需要添加指令，需要在训练文件里为每一条数据添加好生成新的训练文件，然后再加载模型的时候使用设置好的指令。query_instruction_for_retrieval是训练完后使用FlagModel加载模型的参数。
后面我们会完善这一功能，添加到训练模块里。

**Comment by Yyy11181** [comment]
我按照您的要求重新处理了数据。
<img width="665" alt="image" src="https://github.com/FlagOpen/FlagEmbedding/assets/96362397/56f2e2fa-3a4c-4e8a-b24e-56cc4340602d">
之后这样加上了，
<img width="676" alt="image" src="https://github.com/FlagOpen/FlagEmbedding/assets/96362397/2d2785f4-0a24-4b45-a4a9-9300697a7645">
可以吗？

**Comment by Yyy11181** [comment]
query_instruction_for_retrieval是训练完后使用FlagModel加载模型的参数。 这个是什么意思啊? 训练可以加上这个参数吧？

**Comment by Yyy11181** [comment]
您好 我统计了一下数据集中neg 的数量，但是最小和平均之前存在很大的差异 我该选择哪个呢？
<img width="353" alt="image" src="https://github.com/FlagOpen/FlagEmbedding/assets/96362397/5f40c44c-4d36-428d-911e-280c64906196">
再次感谢您

**Comment by staoxiao** [comment]
> 我按照您的要求重新处理了数据。 <img alt="image" width="665" src="https://user-images.githubusercontent.com/96362397/266201598-56f2e2fa-3a4c-4e8a-b24e-56cc4340602d.png"> 之后这样加上了， <img alt="image" width="676" src="https://user-images.githubusercontent.com/96362397/266201744-2d2785f4-0a24-4b45-a4a9-9300697a7645.png"> 可以吗？

可以的，但不需要为在DataAruguments里增加这一参数，并没有用处，训练的没有用到这个参数。推理的时候是用上了的
`model = FlagModel('BAAI/bge-large-zh', query_instruction_for_retrieval="为这个句子生成表示以用于检索相关文章：")
`

**Comment by staoxiao** [comment]
> 您好 我统计了一下数据集中neg 的数量，但是最小和平均之前存在很大的差异 我该选择哪个呢？ <img alt="image" width="353" src="https://user-images.githubusercontent.com/96362397/266203833-5f40c44c-4d36-428d-911e-280c64906196.png"> 再次感谢您

建议就选择平均数量，训练脚本会自动将太少的neg重复几遍以达到预设值。

**Comment by Yyy11181** [comment]
那训练数据加上指令有用吗？训练没用到这个参数



---原始邮件---
发件人: ***@***.***&gt;
发送时间: 2023年9月7日(周四) 下午2:04
收件人: ***@***.***&gt;;
抄送: ***@***.******@***.***&gt;;
主题: Re: [FlagOpen/FlagEmbedding] 关于微调参数“tempeture”等 (Issue #58)




  
我按照您的要求重新处理了数据。  之后这样加上了，  可以吗？
  
可以的，但不需要为在DataAruguments里增加这一参数，并没有用处，训练的没有用到这个参数。推理的时候是用上了的
 model = FlagModel('BAAI/bge-large-zh', query_instruction_for_retrieval="为这个句子生成表示以用于检索相关文章：")

**Comment by Yyy11181** [comment]
好的，感谢您



---原始邮件---
发件人: ***@***.***&gt;
发送时间: 2023年9月7日(周四) 下午2:05
收件人: ***@***.***&gt;;
抄送: ***@***.******@***.***&gt;;
主题: Re: [FlagOpen/FlagEmbedding] 关于微调参数“tempeture”等 (Issue #58)




  
您好 我统计了一下数据集中neg 的数量，但是最小和平均之前存在很大的差异 我该选择哪个呢？  再次感谢您
  
建议就选择平均数量，训练脚本会自动将太少的neg重复几遍以达到预设值。

**Comment by Yyy11181** [comment]
@staoxiao 那训练数据加上指令有用吗？训练没用到这个参数的话

**Comment by staoxiao** [comment]
有的，训练的时候用的训练数据，如果训练数据里已经带上了指令，那么就会和指令一起训练。

**Comment by Yyy11181** [comment]
好 再次感谢您

**Comment by Yyy11181** [comment]
您好 遇到一个很奇怪的问题：
<img width="886" alt="image" src="https://github.com/FlagOpen/FlagEmbedding/assets/96362397/0fe926b0-f0f6-4e55-8717-c1117f74360a">
我这个不是在分布式训练上的参数设置，但是我想用gpu,但是一直显示这个错误，我也加了相应的这2个参数。再次感谢

**Comment by staoxiao** [comment]
不应该添加这两个参数，如果只想使用一张卡，`torchrun --nproc_per_node 1`就可以了

**Comment by staoxiao** [comment]
您好，增加了instuction 功能，现在可以直接通过query_instruction_for_retrieval来设置instruction，不需要手动在数据文件里加指令。

**Comment by Yyy11181** [comment]
> 不应该添加这两个参数，如果只想使用一张卡，`torchrun --nproc_per_node 1`就可以了

但是”torchrun “这个分布式训练的命令吗？我目前没有用到分布式训练

**Comment by Yyy11181** [comment]
> 您好，增加了instuction 功能，现在可以直接通过query_instruction_for_retrieval来设置instruction，不需要手动在数据文件里加指令。

好的 感谢我看一下

**Comment by Yyy11181** [comment]
root@-container-43e51183ae-c3721124:~/FlagEmbedding# torchrun --nproc_per_node  1  -m FlagEmbedding.baai_general_embedding.finetune.run  --output_dir   /root/autodl-tmp/fine_tune_bge-large-zh-noinstruct  --model_name_or_path   /root/autodl-tmp/BAAI/bge-large-zh  --train_data    /root/FlagEmbedding/examples/finetune/merged_filtered_data.json   --learning_rate 1e-5 --fp16 --num_train_epochs 5 --per_device_train_batch_size {large batch size; set 1 for toy data} --normlized True --temperature 0.02 --query_max_len 64 --passage_max_len 256 --train_group_size 2 --negatives_cross_device --query_instruction_for_retrieval "为这个句子生成表示以用于检索相关文章：" 
bash: /root/miniconda3/bin/torchrun: No such file or directory  
您好 为啥报这个错啊

**Comment by staoxiao** [comment]
clone整个项目进行安装，或者pip安装最新版。
这个问题应该跟你的环境有关，可以试试升级torch的版本

**Comment by Yyy11181** [comment]
好的 感谢您

**Comment by Yyy11181** [comment]
<img width="259" alt="image" src="https://github.com/FlagOpen/FlagEmbedding/assets/96362397/58232fcd-8696-4028-9b55-c2ad56717d4f">
您好 ，这是微调后的相关文件，直接放入Langchain-Chatchat 进行embedding吗

**Comment by staoxiao** [comment]
您好，我对Langchain-Chatchat不太熟，看代码的话，直接加载是可以的，但是需要注意模型名字中带有'bge-'才会正确调用，为查询加上指令，否则需要手动为查询加上指令：
https://github.com/chatchat-space/Langchain-Chatchat/blob/master/server/knowledge_base/utils.py#L72

**Comment by Yyy11181** [comment]
好的，感谢您

**Comment by Yyy11181** [comment]
您好 有个问题请教一下 在微调的时候，可以通过看微调loss 曲线是否趋于平缓看是否收敛吗？比如这个图：
<img width="416" alt="05bc66f29bf24c5a3e1d3a9c08b2bed" src="https://github.com/FlagOpen/FlagEmbedding/assets/96362397/a2679ca2-edb0-49fd-ab23-fb41771ae179">
就是没有达到效果，

**Comment by staoxiao** [comment]
> 您好 有个问题请教一下 在微调的时候，可以通过看微调loss 曲线是否趋于平缓看是否收敛吗？比如这个图： <img alt="05bc66f29bf24c5a3e1d3a9c08b2bed" width="416" src="https://user-images.githubusercontent.com/96362397/266787005-a2679ca2-edb0-49fd-ab23-fb41771ae179.png"> 就是没有达到效果，

loss曲线可以在某种程度上可以反映是否收敛。但是如果数据量不大的话，loss一直下降也可能导致过拟合。建议在训练的时候选择合适的checkpoint。

**Comment by Yyy11181** [comment]
再次感谢，那如何在训练的时候选择合适的checkpoint呢？

**Comment by staoxiao** [comment]
需要一份测试数据，选择在测试数据上最好的checkpoint。
目前代码里还没有本地任务的测试，需要自行构建评测代码，可以参考C-MTEB的测试。我们后期会逐渐加上本地数据测试的支持。

**Comment by parkLGW** [comment]
> 您好 有个问题请教一下 在微调的时候，可以通过看微调loss 曲线是否趋于平缓看是否收敛吗？比如这个图： <img alt="05bc66f29bf24c5a3e1d3a9c08b2bed" width="416" src="https://user-images.githubusercontent.com/96362397/266787005-a2679ca2-edb0-49fd-ab23-fb41771ae179.png"> 就是没有达到效果，

请问这张图在哪输出的？我好像找不到微调之后的结果在哪里
